module/VuFind/src/VuFind/I18n/Initializer.php

```diff
@@ -0,0 +1,289 @@
+     * @param StorageInterface $cache         Language cache
+     * @param MvcTranslator    $mvcTranslator Translator
+     * @param ViewModel        $viewModel     ViewModel
+     */
+    public function __construct(
+        Request $request,
+        Config $config,
+        CookieManager $cookies,
+        StorageInterface $cache,
+        MvcTranslator $mvcTranslator,
+        ViewModel $viewModel
+    ) {
+        $language = $this->detectLanguage($config, $request, $cookies);
+
+        $this->translator = $mvcTranslator->getTranslator();
+        $this->loader = $this->translator->getPluginManager()
+            ->get(ExtendedIni::class);
+
+        $this->translator->setLocale($language);
+        $this->translator->setCache($cache);
+
+        $this->loader->setFallbacks($this->parseFallbacks($config));
+
+        $rtlLangs = isset($config->LanguageSettings->rtl_langs)
+            ? array_map(
+                'trim', explode(',', $config->LanguageSettings->rtl_langs)
+            ) : [];
+
+        $viewModel->setVariable('userLang', $language);
+        $viewModel->setVariable('allLangs', $config->Languages);
+        $viewModel->setVariable('rtl', in_array($language, $rtlLangs));
+
+        $this->addBaseDir(APPLICATION_PATH . '/languages');
+        $this->addBaseDir(LOCAL_OVERRIDE_DIR . '/languages');
+    }
+
+    /**
+     * Initializes the translator and loader.
+     *
+     * @return void
+     */
+    public function init()
+    {
+        $cache = $this->translator->getCache();
+        $cacheKey = md5(implode('', $this->baseDirs));
+
+        if (!$cache->hasItem($cacheKey)) {
+            $cache->setItem($cacheKey, $this->detectTextDomains());
+        }
+
+        foreach ($cache->getItem($cacheKey) as $textDomain) {
+            $this->translator->addRemoteTranslations(
+                ExtendedIni::class, $textDomain
+            );
+        }
+
+        $this->loader->setDirs(array_reverse($this->baseDirs));
+    }
+
+    /**
+     * Parses the configured language fallbacks.
+     *
+     * @param Config $config Configuration
+     *
+     * @return array
+     */
+    protected function parseFallbacks(Config $config)
+    {
+        preg_match_all(
+            "#([*a-z-]+):([a-z-]+)#",
+            $config->LanguageSettings->fallbacks,
+            $matches,
+            PREG_SET_ORDER
+        );
+
+        return iterator_to_array(
+            (function () use ($matches) {
+                foreach ($matches as list(, $locale, $fallback)) {
+                    yield $locale => $fallback;
+                }
+            })()
+        );
+    }
+
+    /**
+     * Adds a language directory which will have higher priority
+     * than all directories added before.
+     *
+     * @param string $baseDir Path to directory.
+     *
+     * @return void
+     */
+    public function addBaseDir(string $baseDir)
+    {
+        $this->baseDirs[] = realpath($baseDir);
+    }
```

Comment:
Using this method new language directories may be added - see [`VuFindTheme\Initializer`](https://github.com/vufind-org/vufind/blob/c5ec6f98ac86daef4200b50d3d65bbbd495371d4/module/VuFindTheme/src/VuFindTheme/Initializer.php#L435-L438).

Reply:
Is there any benefit to designing this in a way that allows specific priorities to be set? Forcing directories to always be added in the same order might be limiting to some situations -- but at the same time, if it's good enough for now, we don't need to overengineer it; we can always add an optional second parameter later.

Reply:
Exactly... and default priority could perhaps be `max priority + 1` to ensure a predictable order is followed if nothing is specified.